### ISSUE TITLE
3D Buildings

### ISSUE THREAD

**Issue by keithamoss** [TerriaJS/terriajs#673]
One of the things we miss from Google Maps Engine is its support for streaming 3D buildings into Google Earth. While support was patchy and tied to Google Earth, we still managed to put it to good use with data supplied by the City of Perth.

![screen shot 2015-05-06 at 15 53 22](https://cloud.githubusercontent.com/assets/1672960/7489484/c371ccd8-f409-11e4-95c5-2757272d3a99.png)

Cesium currently supports [converting Collada files to gITF](https://cesiumjs.org/convertmodel.html) and [loading those into a Cesium scene](https://cesiumjs.org/2014/03/03/Cesium-3D-Models-Tutorial/), but ideally users should be able to access either a single service or a pre-prepared catalogue of buildings grouped together in a single UI element.

AGI have just announced their intention to [support for streaming 3D buildings via CZML, gITF, and GeoJSON](http://cesiumjs.org/2015/04/27/3D-Buildings-in-Cesium/), however the technical details remain unclear.

Consider this a bit of a "nice to have, maybe let's monitor this space for a few months" enhancement request.

**Comment by nahidakbar** [comment]
Nice.

Not sure if it is still around but another thing in this space that had a lot of hype (5-6 years ago) is CityGML. Saw some kool demos of modeled cities and different levels of detail etc back in the day.

Not sure if they ever got around to building standards for streaming their models.

**Comment by stevage** [comment]
Yeah, it's still around, but does seem to have stagnated a bit. Will
McIntosh at City of Greater Geelong has been playing around with trying to
export their city's 3D model in it.

**Comment by maxious** [comment]
There's an example of how to import CityGML into PostGIS and then serve buildings over WFS for Cesium https://github.com/Oslandia/cesium-buildings https://github.com/Oslandia/workshop-3d

(I tried to import KML via GDAL to PostGIS and Geoserver and it didn't work, possibly because there was one "model" https://data.gov.au/dataset/central-geelong-3d-massing-model )

**Comment by kring** [comment]
I played around with that Geelong model in Cesium almost a year ago.  I renamed the .kmz to .zip, extracted the Collada (.dae) file and texture from there, used the Cesium tool to convert it to glTF, and then manually loaded and geolocated it in Cesium.  It worked reasonably well.  I just tried again now and the model is too big for the online glTF conversion tool, so we'd have to do the conversion using the offline tool.

It'd be really cool to get this sort of data into National Map, one way or another.

**Comment by vmora** [comment]
Hi, 

I'm trying to integrate cesium-building into TerriaJS. Cesium-buildings fetches 3D objects from a wfs datasource depending on the visible extend/scale. It uses Cesium.QuadTreePrimitive for that. This allows to access massive datasets without killing the performance because you load tiles, not a monolithic model.

My current plan is to add a new CatalogItem that will add/remove the appropriate Cesium.QuadTreePrimitive (with the cesium-building tile provider) to terria.cesium.primitives and let the tile provider handle the rest. I hope that's the appropriate approach.

For those interested, my WIP is [here](https://github.com/vmora/terriajs/tree/cesium-buildings).

Since I plan to PR, I'd be glad to have advices early in my developments. If PRing now (with a draft) is the preferred approach to collaborate, plz just let me know.

Cheers,

**Comment by kring** [comment]
Hi @vmora, this sounds cool!  Your approach - adding a new type of `CatalogItem` - sounds right to me.  Ideally you would provide some sort of support in Leaflet as well.  Show just the building footprints, maybe?

You're welcome to PR any time you'd like me to start taking a look at the code.  Just add a note at the top of the PR that it's still a work in progress, if it is.

Thanks for working on this!

**Comment by Wjrmc** [comment]
This would be fantastic. A way for city models to be displayed like this would have huge benefit.

**Comment by arytonsenna** [comment]
I am trying to load osm 3d Buildings using the following catalog item :

`{
      "name": "Cesium OSM Buildings",
      "type": "3d-tiles",
      "isEnabled":"true",
      "isMappable":"true",
      "isShown":"true",
      "cacheDuration":"1d",
      "ionAssetId": 96188,
      "style": {
          "color": "Boolean(${feature['cesium#color']}) ? color(${feature['cesium#color']}) : rgb(255,255,255)" 
      },
      "colorBlendMode": "HIGHLIGHT"
}`

But unfortunately when i open TerriaJsMap-next at localhost:3001 using gulp watch
i cannot load the osm buildings layer
and it says that -> "This dataset may currently be experiencing issues"
same is true for "cesium-terrain" layer, but i know for a fact that the layers are functional on TerriaJSMap on the previous version. What might be the issue ?
<img width="368" alt="cesium-osm-bug" src="https://user-images.githubusercontent.com/28485076/97778736-3356ab00-1b9f-11eb-8681-63c336ae25b6.PNG">
